spark_client/domain.py

```diff
@@ -0,0 +1,252 @@
+                defaults[k] = os.path.expandvars(v)
+        return defaults
+
+    @classmethod
+    def read(cls, filename: str) -> "PropertyFile":
+        """Read properties file and return a PropertyFile object.
+
+        Args:
+            filename: input filename
+        """
+        try:
+            return PropertyFile(cls._read_property_file_unsafe(filename))
+        except FileNotFoundError as e:
+            raise e
+
+    def write(self, fp: io.TextIOWrapper) -> "PropertyFile":
+        """Write out a property file to disk.
+
+        Args:
+            fp: file pointer to write to
+        """
+        for k, v in self.props.items():
+            line = f"{k}={v.strip()}"
+            fp.write(line + "\n")
+        return self
+
+    def log(self, log_func: Optional[Callable[[str], None]] = None) -> "PropertyFile":
+        """Print a given dictionary to screen.
+
+        Args:
+            log_func: callable to specify another custom printer function. Default uses the class logger with an
+                      INFO level.
+        """
+
+        printer = (lambda msg: self.logger.info(msg)) if log_func is None else log_func
+
+        for k, v in self.props.items():
+            printer(f"{k}={v}")
+        return self
+
+    @classmethod
+    def _parse_options(cls, options_string: Optional[str]) -> Dict:
+        options: Dict[str, str] = dict()
+
+        if not options_string:
+            return options
+
+        # cleanup quotes
+        line = options_string.strip().replace("'", "").replace('"', "")
+        for arg in line.split("-D")[1:]:
+            kv = arg.split("=")
+            options[kv[0].strip()] = kv[1].strip()
+
+        return options
+
+    @property
+    def options(self) -> Dict[str, Dict]:
+        """Extract properties which are known to be options-like requiring special parsing."""
+        return {
+            k: self._parse_options(v)
+            for k, v in self.props.items()
+            if self._is_property_with_options(k)
+        }
+
+    @staticmethod
+    def _construct_options_string(options: Dict) -> str:
+        result = ""
+        for k in options:
+            v = options[k]
+            result += f" -D{k}={v}"
+        return result
+
+    @classmethod
+    def empty(cls) -> "PropertyFile":
+        """Return an empty property file object."""
+        return PropertyFile(dict())
+
+    def __add__(self, other: "PropertyFile"):
+        return self.union([other])
+
+    def union(self, others: List["PropertyFile"]) -> "PropertyFile":
+        """Merge multiple PropertyFile objects, with right to left priority.
+
+        Args:
+            others: List of Property file to be merged.
+        """
+        all_together = [self] + others
+
+        simple_properties = union(*[prop.props for prop in all_together])
+        merged_options = {
+            k: self._construct_options_string(v)
+            for k, v in union(*[prop.options for prop in all_together]).items()
```

Comment:
not very legible. would like to test the options override bit.

Reply:
Yes, let's do this when doing unittesting. This should be a great use-case for implementing different tests